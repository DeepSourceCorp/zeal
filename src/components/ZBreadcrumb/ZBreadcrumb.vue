<template>
  <div aria-label="Breadcrumb" role="navigation" class="flex space-x-2">
    <slot></slot>
  </div>
</template>

<script>
export default {
  name: 'ZBreadcrumb',
  props: {
    separator: {
      type: String,
      default: '/'
    }
  },
  data() {
    return {
      items: [],
      length: 0
    }
  },
  mounted() {
    this.recalculateItems()
  },
  methods: {
    recalculateItems() {
      this.items = this.$children.filter((child) => child.$options.name === 'ZBreadcrumbItem')
      this.items.map((item, index) => {
        item.index = index
      })
    }
  },
  watch: {
    length() {
      this.recalculateItems()
    }
  }
}
</script>
