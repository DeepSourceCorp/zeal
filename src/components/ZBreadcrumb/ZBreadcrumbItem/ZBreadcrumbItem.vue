<template>
  <span
    class="space-x-2"
    :class="{
      'text-slate': !current
    }"
  >
    <span v-if="this.index !== 0" role="presentation">{{ separator }}</span>
    <span ref="link" role="link">
      <slot></slot>
    </span>
  </span>
</template>

<script>
export default {
  name: 'ZBreadcrumbItem',
  props: {
    isCurrent: {
      default: false,
      type: Boolean
    }
  },
  data() {
    return {
      separator: '',
      index: 0
    }
  },
  computed: {
    current() {
      return this.isCurrent || this.index === this.$parent.length - 1
    }
  },
  mounted() {
    if (this.$parent?.$options.name === 'ZBreadcrumb') {
      this.separator = this.$parent.separator
      this.$parent.length += 1
    }
  },
  beforeDestroy() {
    if (this.$parent?.$options.name === 'ZBreadcrumb') this.$parent.length -= 1
  }
}
</script>
