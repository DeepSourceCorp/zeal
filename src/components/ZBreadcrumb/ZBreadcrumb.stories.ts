import '../../assets/css/tailwind.css'
import '../../assets/css/typography.css'
import '../../assets/css/layout.css'
import ZBreadcrumb from './ZBreadcrumb.vue'
import ZBreadcrumbItem from './ZBreadcrumbItem/ZBreadcrumbItem.vue'
import ZNumberInput from '../ZNumberInput/ZNumberInput.vue'

export default {
  title: 'Breadcrumb',
  component: ZBreadcrumb,
  excludeStories: /.*Data$/
}

export const BasicBreadcrumb = () => ({
  components: { ZBreadcrumb, ZBreadcrumbItem, ZNumberInput },
  data() {
    return {
      count: 3
    }
  },
  template: `<div class='padded-container text-vanilla-100'>
    <z-breadcrumb separator="/" class="text-sm text-vanilla-100">
      <z-breadcrumb-item><a href="#">Autofix</a></z-breadcrumb-item>
      <z-breadcrumb-item>Method has no argument</z-breadcrumb-item>
    </z-breadcrumb>
    <br/>
    <z-breadcrumb separator="/" class="text-sm text-vanilla-100">
      <z-breadcrumb-item><a href="#">Elements</a></z-breadcrumb-item>
      <z-breadcrumb-item v-for="ii in count" :key="ii"><a href="#">Element {{ii}}</a></z-breadcrumb-item>
    </z-breadcrumb>
    <br/>
    <div class="flex gap-x-2">
      Add elements: <z-number-input v-model="count" min="1" class="w-28" />  
    </div>
  </div>`
})

export const BreadcrumbWithCurrentPage = () => ({
  components: { ZBreadcrumb, ZBreadcrumbItem },
  template: `<div class='padded-container'>
  <z-breadcrumb separator="/" class="text-sm text-vanilla-100">
    <z-breadcrumb-item><a href="#">Home</a></z-breadcrumb-item>
    <z-breadcrumb-item><a href="#">Repository</a></z-breadcrumb-item>
    <z-breadcrumb-item isCurrent>Settings</z-breadcrumb-item>
  </z-breadcrumb>
    </div>`
})
