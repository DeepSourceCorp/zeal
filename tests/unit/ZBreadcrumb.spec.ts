import ZBreadcrumb from '../../src/components/ZBreadcrumb'
import ZBreadcrumbItem from '../../src/components/ZBreadcrumb/ZBreadcrumbItem'

import { mount } from '@vue/test-utils'

const BasicBreadcrumb = {
  template: `<z-breadcrumb separator="/" class="text-sm text-vanilla-100">
    <z-breadcrumb-item><a href="#">Autofix</a></z-breadcrumb-item>
    <z-breadcrumb-item>Method has no argument</z-breadcrumb-item>
  </z-breadcrumb>`,
  components: {
    ZBreadcrumb,
    ZBreadcrumbItem
  }
}

const BreadcrumbWithCurrentPage = {
  template: `<z-breadcrumb separator="/" class="text-sm text-vanilla-100">
    <z-breadcrumb-item><a href="#">Home</a></z-breadcrumb-item>
    <z-breadcrumb-item><a href="#">Repository</a></z-breadcrumb-item>
    <z-breadcrumb-item isCurrent>Settings</z-breadcrumb-item>
  </z-breadcrumb>`,
  components: {
    ZBreadcrumb,
    ZBreadcrumbItem
  }
}

describe('ZBreadcrumb', () => {
  it('renders a default breadcrumb component', () => {
    let wrapper = mount(BasicBreadcrumb)
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('renders a breadcrumb component with current set by the user', () => {
    let wrapper = mount(BreadcrumbWithCurrentPage)
    expect(wrapper.html()).toMatchSnapshot()
  })
})
