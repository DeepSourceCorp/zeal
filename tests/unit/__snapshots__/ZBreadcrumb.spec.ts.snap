// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ZBreadcrumb renders a breadcrumb component with current set by the user 1`] = `<div aria-label="Breadcrumb" role="navigation" class="flex space-x-2 text-sm text-vanilla-100"><span class="space-x-2 text-slate"><!----> <span role="link"><a href="#">Home</a></span></span> <span class="space-x-2 text-slate"><!----> <span role="link"><a href="#">Repository</a></span></span> <span class="space-x-2"><!----> <span role="link">Settings</span></span></div>`;

exports[`ZBreadcrumb renders a default breadcrumb component 1`] = `<div aria-label="Breadcrumb" role="navigation" class="flex space-x-2 text-sm text-vanilla-100"><span class="space-x-2 text-slate"><!----> <span role="link"><a href="#">Autofix</a></span></span> <span class="space-x-2 text-slate"><!----> <span role="link">Method has no argument</span></span></div>`;
